### DISCUSSION TITLE
Adding MQTT Discovery for Home Assistant

### DISCUSSION THREAD

**Discussion by r0bb10** [dumpfheimer/olimpia_splendid_bi2_modbus_controller#8] [Ideas]
Hi man! Take a look at this configuration below, this should fix the Home Assistant configuration with `sendHomeAssistantConfiguration()` by using the correct MQTT Auto Discovery function (details [here](https://www.home-assistant.io/integrations/mqtt/#mqtt-discovery)) that i have tested manually by publishing a JSON in the correct homeassistant topic.

The whole `sendHomeAssistantConfiguration()` should be replaced by something like `sendAutoDiscoveryMsg()`, i tried to write it correctly here:

```
void sendAutoDiscoveryMsg() {
  DynamicJsonDocument doc(1024);
  char buffer[256];

  doc["name"] = "Fancoil " + addr + "";
  doc["unique_id"] = "hvac_" + clientId + "-" + addr + "";
  doc["availability_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/state";
  doc["payload_available"] = "online";
  doc["payload_not_available"] = "offline";
  doc["mode_command_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/mode/set";
  doc["mode_state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/mode/state";
  doc["modes"] = "heat", "cool", "off";
  doc["min_temp"] = "15";
  doc["max_temp"] = "30";
  doc["precision"] = "0.5";
  doc["retain"] = "true";
  doc["temperature_command_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/setpoint/set";
  doc["temperature_state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/setpoint/state";
  doc["temp_step"] = "0.5";
  doc["fan_mode_command_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/fan_speed/set";
  doc["fan_mode_state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/fan_speed/state";
  doc["fan_modes"] = "auto", "high", "low", "night";
  JsonObject device  = doc.createNestedObject("device");
  device["name"] = "Fancoil " + addr + "";
  device["via_device"] = "Fancoil CTRL";
  device["identifiers"] = "hvac_" + clientId + "-" + addr + "";
  device["model"] = "B0872 ModBus";
  device["manufacturer"] = "Olimpia Splendid";
  device["model"] = "Olimpia Splendid";
  device["sw_version"] = "1.0";
  serializeJson(doc, buffer);
  client.publish("homeassistant/climate/hvac_" + clientId + "-" + addr + "/config", buffer);

#ifdef LOAD_WATER_TEMP
  doc["name"] = "Water Sensor";
  doc["unique_id"] = "hvac_" + clientId + "-" + addr + "_water_sensor";
  doc["device_class"] = "temperature";
  doc["state_class"] = "measurement";
  doc["unit_of_measurement"] = "C";
  doc["state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/water_sensor/state";
  serializeJson(doc, buffer);
  client.publish("homeassistant/sensor/hvac_" + clientId + "-" + addr + "_water_sensor/config", buffer);
#endif

#ifdef LOAD_AMBIENT_TEMP
  doc["name"] = "Ambient Sensor";
  doc["unique_id"] = "hvac_" + clientId + "-" + addr + "_ambient_sensor";
  doc["device_class"] = "temperature";
  doc["state_class"] = "measurement";
  doc["unit_of_measurement"] = "C";
  doc["state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/ambient_sensor/state";
  serializeJson(doc, buffer);
  client.publish("homeassistant/sensor/hvac_" + clientId + "-" + addr + "_ambient_sensor/config", buffer);
#endif
}
```

The discovery should auto-populate the correct entity in the MQTT integration and should have the ambient/water sensor also if defined, i'm not sure about how many fancoils should be pushed tho, i think it should publish the registered ones if did not mess up anything. The format is JSON that has to be serialized before it can be published in the correct topic but looking online it should work as is. Manually sending the JSON works as expected, this is an example:

```
{
	"name": "Fancoil 1",
	"send_if_off": true,
	"unique_id": "hvac_C8-C9-A3-54-1B-C7-1",
	"availability_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/state/state",
	"payload_available": "online",
	"payload_not_available": "offline",
	"current_temperature_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/ambient_temperature/state",
	"fan_mode_command_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/fan_speed/set",
	"fan_mode_state_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/fan_speed/state",
	"fan_modes": ["auto", "high", "low", "night"],
	"mode_command_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/mode/set",
	"mode_state_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/mode/state",
	"modes": ["heat", "cool", "off"],
	"min_temp": 15,
	"max_temp": 30,
	"precision": 0.5,
	"retain": true,
	"temperature_command_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/setpoint/set",
	"temperature_state_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/1/setpoint/state",
	"temp_step": 0.5,
	"device": {
		"manufacturer": "Olimpia Splendid",
		"via_device": "Fancoil CTRL",
		"identifiers": "hvac_C8-C9-A3-54-1B-C7-1",
		"name": "Fancoil 1",
		"model": "B0872 ModBus",
		"sw_version": "1.0"
	}
}
```

![image](https://github.com/dumpfheimer/olimpia_splendid_bi2_modbus_controller/assets/16479888/61bc40e0-5e36-4597-9a6f-ea48cbc01570)
![image](https://github.com/dumpfheimer/olimpia_splendid_bi2_modbus_controller/assets/16479888/3cc7cce2-92a6-4a39-902c-50eefc5f889e)

I did not make a push request because i'm not sure how to implement and remove the old parts, but this should completely avoid to manually create mqtt.yaml entity and find out the correct topics in the broker, polishing the deploy.

Cheers!

**Comment by dumpfheimer** [comment]
Hi there!

Sorry for the late reply.
This looks quite nice! Would you be able to open a pull request?

Is this using the ArduinoJson library?

Thanks for your effort.

**Comment by r0bb10** [comment]
Hi!

So those below are the JSON sent via mqtt discovery (each has to be sent in a topic called "homeassistant/climate/NAME/config" - for the climate entity - and "homeassistant/sensor/NAME/config" - for the sensor entity), the arduino version is using the ArduinoJson library to serialize and have this published but there are some problems.

Ok first of all i dont know how to code it correctly, specifically have the arduino code "know" the fancoil address to publish in each config topic, but that is probably easy for you to understand. I "think" the document is correct with all the doc part necessary present, but i'm not really sure.

The problem is that the config subtopic in the homeassistant topic is volatile and does not survive a homeassistant restart, even if the mqtt broker does not reboot.. for example i use hassio supervised and homeassistant core and mosquitto are 2 different containers, on update just core reboots and mosquitto stays up but even so the homeassistant topic is cleared.
So the AutoDiscovery has to be tied to homeassistant somehow, and the config message has to be resent.
Now `sendHomeAssistantConfiguration()` (which i noticed is needed in the code so cannot be removed) is sent to mosquitto on mqtt reconnect but that would not work if homeassistant restarts for whatever reason (update, crash, manual restart) so the arduino code should push `sendAutoDiscoveryMsg()` in a different way.. time based? every 5 min? mm not clean i guess, somehow the loop should know when hass needs the config and when to push it. I have other mqtt autodiscovered entities that do that correctly (namely Valetudo for the vacuum entity does it right), the topic is cleared but it shows up after a little while.

I did not add the `LOAD_AMBIENT_TEMP` since i noticed that in rE mode that value in register 0 is a mirrored value of register 103 is it's ignoring the onboard sensor. Need more confirmation on that tho.


```
void sendAutoDiscoveryMsg() {
  DynamicJsonDocument doc(1024);
  char buffer[256];

  doc["name"] = "Fancoil " + addr + "";
  doc["icon"] = "mdi:home-thermometer-outline";
  doc["send_if_off"] = "true";
  doc["unique_id"] = "hvac_" + clientId + "-" + addr + "";
  doc["availability_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/state";
  doc["payload_available"] = "online";
  doc["payload_not_available"] = "offline";
  doc["mode_command_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/mode/set";
  doc["mode_state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/mode/state";
  doc["action_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/action/state";
  doc["modes"] = "heat", "cool", "off";
  doc["min_temp"] = "15";
  doc["max_temp"] = "30";
  doc["precision"] = "0.1";
  doc["retain"] = "true";
  doc["temperature_command_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/setpoint/set";
  doc["temperature_state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/setpoint/state";
  doc["temp_step"] = "0.5";
  doc["fan_mode_command_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/fan_speed/set";
  doc["fan_mode_state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/fan_speed/state";
  doc["fan_modes"] = "auto", "high", "low", "night";
  JsonObject device  = doc.createNestedObject("device");
  device["name"] = "Fancoil " + addr + "";
  device["via_device"] = "Fancoil CTRL";
  device["identifiers"] = "hvac_" + clientId + "-" + addr + "";
  device["model"] = "Bi2 SL Smart 400";
  device["manufacturer"] = "Olimpia Splendid";
  serializeJson(doc, buffer);
  client.publish("homeassistant/climate/hvac_" + clientId + "-" + addr + "/config", buffer);

#ifdef LOAD_WATER_TEMP
  doc["name"] = "Water Sensor";
  doc["unique_id"] = "watersensor_" + clientId + "-" + addr + "";
  doc["send_if_off"] = "true";
  doc["device_class"] = "temperature";
  doc["state_class"] = "measurement";
  doc["unit_of_measurement"] = "°C";
  doc["suggested_display_precision"] = 0;
  doc["state_topic"] = "fancoil_ctrl/" + clientId + "/" + addr + "/water_sensor/state";
  serializeJson(doc, buffer);
  client.publish("homeassistant/sensor/watersensor_" + clientId + "-" + addr + "/config", buffer);
#endif
```

![image](https://github.com/dumpfheimer/olimpia_splendid_bi2_modbus_controller/assets/16479888/cd7565e4-353a-4883-805a-fc6da985c444)


```
{
  "name": "HVAC",
  "icon": "mdi:home-thermometer-outline",
  "send_if_off": true,
  "unique_id": "hvac_C8-C9-A3-54-1B-C7-5",
  "availability_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/state/state",
  "payload_available": "online",
  "payload_not_available": "offline",
  "current_temperature_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/ambient_temperature/state",
  "fan_mode_command_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/fan_speed/set",
  "fan_mode_state_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/fan_speed/state",
  "fan_modes": [
    "auto",
    "high",
    "low",
    "night"
  ],
  "action_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/action/state",
  "mode_command_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/mode/set",
  "mode_state_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/mode/state",
  "modes": [
    "heat",
    "cool",
    "off"
  ],
  "min_temp": 15,
  "max_temp": 30,
  "precision": 0.1,
  "retain": true,
  "temperature_command_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/setpoint/set",
  "temperature_state_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/setpoint/state",
  "temp_step": 0.5,
  "device": {
    "manufacturer": "Olimpia Splendid",
    "via_device": "Fancoil CTRL",
    "identifiers": "hvac_C8-C9-A3-54-1B-C7-5",
    "name": "Fancoil Studio",
    "model": "Bi2 SL Smart 400"
  }
}

{
  "name": "Water Sensor",
  "unique_id": "watersensor_C8-C9-A3-54-1B-C7-5",
  "send_if_off": true,
  "device_class": "temperature",
  "state_topic": "fancoil_ctrl/C8-C9-A3-54-1B-C7/5/water_sensor/state",
  "unit_of_measurement": "°C",
  "suggested_display_precision": 0,
  "device": {
    "manufacturer": "Olimpia Splendid",
    "via_device": "Fancoil CTRL",
    "identifiers": "hvac_C8-C9-A3-54-1B-C7-5",
    "name": "Fancoil Studio",
    "model": "Bi2 SL Smart 400"
  }
}

**Comment by dumpfheimer** [comment]
Hey Roberto,
Sorry for the late reply. What is the state of this? Have you worked on this some more? Do you have (or is this) a stable version of the auto discovery?

**Comment by materoxxx** [comment]
Hi, I Just set up the first device with the help of dumpfheimer.  In HA in mqtt I see the device but I'm unable to set temperature, If I can get it to work like yours it'll be a dream. Thanks

**Comment by r0bb10** [comment]
> Hey Roberto, Sorry for the late reply. What is the state of this? Have you worked on this some more? Do you have (or is this) a stable version of the auto discovery?

Hi!

uhm actually i'm using the same code i posted above (just serialized using codebeautify.org) as a payload in an automation inside homeassistant, and it works great.
Every time i reboot homeassistant for whatever reason the automation triggers and publishes the discovery inside itself, so it's not really an autodiscovery function in fancoilctrl.

```
alias: HVAC Discovery
description: ""
triggers:
  - event: start
    trigger: homeassistant
    enabled: true
conditions: []
  - alias: Expose Studio HVAC
    data:
      qos: 0
      retain: false
      topic: homeassistant/climate/fancoil_studio/config
      payload: >-
        {   "name":"HVAC",   "icon":"mdi:home-thermometer-outline",  
        "send_if_off":true,   "unique_id":"hvac_C8-C9-A3-54-1B-C7",  
        "availability_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/state/state",  
        "payload_available":"online",   "payload_not_available":"offline",  
        "current_temperature_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/ambient_temperature/state",  
        "fan_mode_command_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/fan_speed/set",  
        "fan_mode_state_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/fan_speed/state",  
        "fan_modes": [     "auto",     "high",     "low",     "night"   ],  
        "action_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/action/state",  
        "mode_command_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/mode/set",  
        "mode_state_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/mode/state",  
        "modes": [     "heat",     "cool",     "off"   ],   "min_temp":15,  
        "max_temp":30,   "precision":0.1,   "retain":true,  
        "temperature_command_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/setpoint/set",  
        "temperature_state_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/setpoint/state",  
        "temp_step":0.5,   "device": {     "manufacturer":"Olimpia
        Splendid",     "via_device":"Fancoil CTRL",    
        "identifiers":"hvac_C8-C9-A3-54-1B-C7",     "name":"Fancoil Studio",    
        "model":"Bi2 SL Smart 400"   } }
    action: mqtt.publish
  - alias: Expose Studio Water Sensor
    data:
      qos: 0
      retain: false
      topic: homeassistant/sensor/fancoil_studio_water_temp/config
      payload: >-
        {   "name":"Water Sensor",  
        "unique_id":"watersensor_C8-C9-A3-54-1B-C7",   "send_if_off":true,  
        "device_class":"temperature",  
        "state_topic":"fancoil_ctrl/C8-C9-A3-54-1B-C7/5/water_sensor/state",  
        "unit_of_measurement":"°C",   "suggested_display_precision":0,  
        "device": {     "manufacturer": "Olimpia Splendid",     "via_device":
        "Fancoil CTRL",     "identifiers": "hvac_C8-C9-A3-54-1B-C7",     "name":
        "Fancoil Studio",     "model": "Bi2 SL Smart 400"   } }
    action: mqtt.publish
```
Running this automation loads the mqtt device and creates the two entity (the hvac/climate and the extra water sensor).

Not having homeassistant do this would be ideal (my workaround just voids to manually create an mqtt entity) so this payload has to come from fancoilctrl.. as a practical example Zigbee2Mqtt has autodiscovery function and when a new device is paired it makes it visible inside homeassistant but i dont know what the logic could be in fancoilctrl.

There should be probably some kind of template based on what i've pasted that gets populated with actual data (some real line the mac from the master, and the ids from the registered slaves) but some data can be hardcoded like the manufacturer and model.

Any idea on how to implement this?

**Comment by dumpfheimer** [comment]
I started working on this but was not able to test it because my setup does not run over home assistant. But the climate entity does show up now. Will need some more work though

**Comment by r0bb10** [comment]
> I started working on this but was not able to test it because my setup does not run over home assistant. But the climate entity does show up now. Will need some more work though

Oh wow nice i didn't notice it!
Last time i tried to compile the latest version i had troubles with platformio doing it so i just went back to arduinoide and the last known version.. i will retry cloning the new version and compile it to test, hopefully report back soon :)

**Comment by materoxxx** [comment]
R0bb10 , If I want to emulate what you did, should I write your config above in the config.yaml ? changing of course the fancoil name to match mine etc. Because right now I see my fancoil like this: https://imgur.com/a/mru4pfZ and I can't change the values from mqtt only using http in the webpage as the readme says. Thank you

**Comment by r0bb10** [comment]
> R0bb10 , If I want to emulate what you did, should I write your config above in the config.yaml ? changing of course the fancoil name to match mine etc. Because right now I see my fancoil like this: https://imgur.com/a/mru4pfZ and I can't change the values from mqtt only using http in the webpage as the readme says. Thank you

@materoxxx as the image you posted you did not create a climate entity in homeassistant but importend the single sensors as mqtt, you have to create a mqtt climate described here [https://www.home-assistant.io/integrations/climate.mqtt/](url)

you can try my automation, i made one for your specific esp to try it out:

```
description: ""
mode: single
triggers:
  - event: start
    trigger: homeassistant
    enabled: true
conditions: []
actions:
  - alias: Expose HVAC
    data:
      qos: 0
      retain: false
      payload: >-
        {   "name":"HVAC",   "icon":"mdi:home-thermometer-outline",  
        "send_if_off":true,   "unique_id":"hvac_C4-B8-D5-2C-C2-98-1",  
        "availability_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/state/state",  
        "payload_available":"online",   "payload_not_available":"offline",  
        "current_temperature_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/ambient_temperature/state",  
        "fan_mode_command_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/fan_speed/set",  
        "fan_mode_state_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/fan_speed/state",  
        "fan_modes": [     "auto",     "high",     "low",     "night"   ],  
        "action_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/action/state",  
        "mode_command_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/mode/set",  
        "mode_state_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/mode/state",  
        "modes": [     "heat",     "cool",     "off"   ],   "min_temp":15,  
        "max_temp":30,   "precision":0.1,   "retain":true,  
        "temperature_command_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/setpoint/set",  
        "temperature_state_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/setpoint/state",  
        "temp_step":0.5,   "device": {     "manufacturer":"Olimpia
        Splendid",     "via_device":"Fancoil CTRL",    
        "identifiers":"hvac_C4-B8-D5-2C-C2-98-1",     "name":"Fancoil 1",    
        "model":"Bi2 SL Smart 400"   } }
      topic: homeassistant/climate/fancoil_1/config
    enabled: true
    action: mqtt.publish
  - alias: Expose HVAC Water Sensor
    data:
      qos: 0
      retain: false
      topic: homeassistant/sensor/fancoil_1_water_temp/config
      payload: >-
        {   "name":"Water Sensor",  
        "unique_id":"watersensor_C4-B8-D5-2C-C2-98-1",   "send_if_off":true,  
        "device_class":"temperature",  
        "state_topic":"fancoil_ctrl/C4-B8-D5-2C-C2-98/1/water_sensor/state",  
        "unit_of_measurement":"°C",   "suggested_display_precision":0,  
        "device": {     "manufacturer": "Olimpia Splendid",     "via_device":
        "Fancoil CTRL",     "identifiers": "hvac_C4-B8-D5-2C-C2-98-1",    
        "name": "Fancoil 1",     "model": "Bi2 SL Smart 400"   } }
    enabled: true
    action: mqtt.publish
```

copy this whole code, go in automation -> create new -> on the 3 dots menu select "edit in yaml" remove everything and paste this code -> save the automation.

once saved without restarting homeassistant just press run, the fancoil should appear in the mqtt devices as a climate entity correctly.

**Comment by materoxxx** [comment]
Thank you very much, in the meantime The code was updated from the author and I now see it as a climate entity without the automation. The only small issue is the ambient temperature, maybe with your automation I can assign a real external temp sensor I have in the room to the climate entity. Because right now I don't see in config.yaml or anywhere how to edit the climate entity that was created by mqtt.

**Comment by materoxxx** [comment]
As I imagined the automation is not working (not available) because the device now already appears correctly as a climate entity with the latest changes to the code, thank you anyway, I'm going to try an automation to set ambient temp the same as a different sensor I have, the only issue I see is that the precision I see in the climate entity is 1 instead of .1 or .5, so I have to see how it plays

**Comment by r0bb10** [comment]
@dumpfheimer i tested the latest changes and discovery works correctly! i saw that model manufacturer and via_device are commented out (they report as unknow/unknowin the pic attatched), and indedd seems that via_device it's not needed to add in the same device the water temp.

all the device registered are showing up (6 in my case) and also the controller itself.

![image](https://github.com/user-attachments/assets/5b791ac0-7281-496a-8c3f-bafc857482bf)

here a comparison between the same two devices, the first selfdiscovered via the automation and the second autodiscovered from the controller

![image](https://github.com/user-attachments/assets/d46a462c-cbf5-474a-99e7-a28ea53bf29d)

![image](https://github.com/user-attachments/assets/167c97ca-f3bf-4d6e-833c-3d7df9cd9eb1)

it works and it works good, there should only be "ui" improvement, let me explain with an example for another device i have

![image](https://github.com/user-attachments/assets/bb3d8784-d3b3-49ac-86a6-9f7cd087d068)

in the Control 
- the select mode and fanspeed are already in the climate entity so redundant
- the ambient number should have decimal with .1 precision
- swing should be visible only if the fancoil supports it (that could be tricky to filter)

in the Sensor
- the water sensor should not have decimals
- setpoint is redundant

in the Diagnostic (dunno how to add that) should have only information data
- shoud have is_consuming (that is now in sensor)
- should have state online (that is now in sensor)

Cheers, great job as usual!
